### ISSUE TITLE
Request: currying `numpy.einsum` for repeated calls with the same subscript string?

### ISSUE THREAD

**Issue by congma** [numpy/numpy#6631]
Hi,

During my experience with the `numpy.einsum` function, I discovered that I frequently use it in the following pattern:  The zeroth argument to `einsum()`, the string describing the subscripts, is a fixed one, but the array arguments will vary during runtime.  That is to say, I apply the same summation rule to many different array arguments repeatedly.

I don't know if `einsum()` keeps its subscript-parsing results in a cache, but if not, this could be a bit wasteful, because the same subscript description need to be unnecessarily parsed again and again.

So I wonder if it is possible to have a "curriable" version of `einsum` that, when called with the subscript description string, returns a callable that has already parsed the string.  This returned callable is then ready to be applied to the array arguments.

On the other hand, I'm not sure about the performance impact of parsing.  Is this kind of optimization worth the effort?

Just my 2c.

**Comment by charris** [comment]
Might be faster, although I expect a good bit of the savings would be in converting the python string to a C string, especially in Python 3. Einsum does seem faster when called directly from C, but I haven't done actual benchmarking. Saving the parsed result would be handy, but would require some work and a way to expose the result in python.

**Comment by njsmith** [comment]
Yeah, it would make sense to have an interface like this, and it would allow for some nice optimizations by amortizing the parsing etc. (Even more so if we started getting cleverer about optimizing einsum execution.) So whenever someone gets around to implementing that kind of amortization, then I agree, this would be a good interface for it :-)
